server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -390,52 +391,35 @@ protected void createMissingIndicesAndIndexData(
         long startTime
     ) {
         final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
+        // Optimizing when there are no prerequisite actions
         if (autoCreateIndices.isEmpty()) {
             executeBulk(task, bulkRequest, startTime, listener, executorName, responses, indicesThatCannotBeCreated);
-        } else {
-            final AtomicInteger counter = new AtomicInteger(autoCreateIndices.size());
+            return;
+        }
+        Runnable executeBulkRunnable = () -> threadPool.executor(executorName).execute(new ActionRunnable<>(listener) {
+            @Override
+            protected void doRun() {
+                executeBulk(task, bulkRequest, startTime, listener, executorName, responses, indicesThatCannotBeCreated);
+            }
+        });
+        try (RefCountingRunnable refs = new RefCountingRunnable(executeBulkRunnable)) {
             for (String index : autoCreateIndices) {
-                createIndex(index, bulkRequest.timeout(), new ActionListener<>() {
-                    @Override
-                    public void onResponse(CreateIndexResponse result) {
-                        if (counter.decrementAndGet() == 0) {
-                            forkExecuteBulk(listener);
+                createIndex(index, bulkRequest.timeout(), refs.acquireListener().delegateResponse((l, e) -> {
```

Comment:
Chaining `delegateResponse` and `map` in one expression is kind of suspicious, you're effectively overriding both `onResponse` and `onFailure` at the same time. Maybe just create a `new ActionListener` directly, with something like `ActionListener.releaseAfter(new ActionListener<>{...}, refs.acquire())`.

Reply:
Thank you @DaveCTurner ! Good point, let me try it.